### ISSUE TITLE
axmls date/datetime different mysql vs sql server vs oracle

### ISSUE THREAD

**Issue by eharris200** [ADOdb/ADOdb#283]
adodb 5.20.7,  windows 7, sql server 2012, iis 7.5, oracle 12.1.0.2, mysql 5.7

If I pick type of D, mysql gets Date, sql server gets date time, oracle gets date
If i pick type of T, mysql get DateTime, sql server gets time(7), oracle gets date

Is there an option I can pick to get mysql => datetime, sql server => datetime, and oracle => date ?

One work around I see is that I create all my tables with type D.  As I'm looping through in my creation script, if I'm creating tables for mysql, replace all 'D' type with 'T'.  Hoping I missed something in doc that would allow me to define in .xml table definition files.

**Comment by dregad** [comment]
I don't use the XML schema myself, so I'm not able to answer your question, sorry. Maybe @mnewnham has an idea.

**Comment by peterdd** [comment]
tl;dr:  time(7) of microsoft sql server for 'T' seems wrong to me, should be DATETIME.


And 'D' should be DATE too, but also the access and sybase datadict drivers store that as DATETIME, oracle as DATE (but contains time data too)

Well, the underlying problem there is that there are only 2 main abbreviations ('T' and 'D') for that date/time column stuff where probably 3 are needed:

1. when just storing a date required (2017-07-17)  - currently the 'D' in adodb datadict
2. when storing just a time value like (13:07:59 .12345) is required - not defined yet in adodb?
3. for date and time (2017-07-17 13:07:59 .12345) - currently the 'T' in adodb datadict

If we would have that 3 options, every database driver could easy choose the right column type without overlapping with other database driver solutions.

If I interprete the incomplete documentation of the adodb datadict stuff right, 'T' (or its probable alias 'TS')
shall be a column that contains info of date **and** time.

https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetime-transact-sql
https://docs.microsoft.com/en-us/sql/t-sql/data-types/time-transact-sql

You are using the mssqlnative driver?


Thoughts:
The 'T' type within the adodb datadict related classes should definitely better documented that it must be a column type in each database type that contains date and also time information.

It is a bit weired because of the inconsistency namings across the databases:

Some database types use **date** for only date (mysql, postgresql), some store date and time there too(oracle).

Some database types use **timestamp** for date and time(postgresql).

'D' and 'T' interpretation seems a bit ambiguous and messed across the different datadict db driver specific overwrites in datadict/datadict-*.php files:


See adodb/adodb-datadict.inc.php function MetaType() and ActualType()
and 
```sh
cd adodb/datadict;
grep -1 "'T':" datadict-*.php
```

https://docs.oracle.com/cd/B28359_01/server.111/b28298/ch4datetime.htm#i1006006 tells:

> Oracle Database DATE columns always contain fields for both **date and time**.

The 'TS' is also not yet documented (neither source nor adodb.org) and probably just an alias of 'T'.

**Comment by DavidBruchmann** [comment]
If someone is fixing anything, consider also this one: https://github.com/ADOdb/ADOdb/issues/144

**Comment by eharris200** [comment]
I am currently using the native sql server driver.  3 types would work for me.  Right now my work around is I store all XML files with 'D' and when the tables are pulled in a call ParseSchemaFile and if it returns DATE and I'm installing against MYSQL I turn this into DATETIME.

**Comment by mnewnham** [comment]
There has always been a problem with the SQL server driver because in earlier releases, (pre 2005 I think), there was no DATE type field, so the D type was mapped by necessity to the DATETIME field, which has caused a lot of compatibility issues, plus vagueness and ambiguity in the documentation. 

Up until now, there has been no real attempt to fix this problem, given that the DATE field is now well established in SQL Server.

I'm currently in a discussion with @dregad regarding:

1. Standardizing the MetaTypes across all databases
2. Implementing constants to replace 'D",'T','X' etc
3. Adding methods that allow the addition of user defined MetaTypes/ActualTypes, as well as user defined overrides for existing types 

The new usage is explained in the documentation page [setActualType()](http://adodb.org/dokuwiki/doku.php?id=v5:dictionary:setactualtypel)

Hopefully this will allow us to move forward with a much more clearly defined MetaType/ActualType definition, and we can drop the ambiguity.

**Comment by clarumedia** [comment]
Wondering if anything came of those conversations with Damien Regad ?

For my money:

adodb-mysqli.inc.php 
function MetaType

Surely 

- MySQL data type DATETIME should return "D" 
- MySQL data type TIME should return "T" 

?
